Failure with CachingOptimizer

This seems like a bug in `CachingOptimizer`
```julia
using Test
import MathOptInterface as MOI
import CSDP

function tests(cache)
    model = MOI.instantiate(CSDP.Optimizer, with_bridge_type = Float64)
    if cache
        model = MOI.Utilities.CachingOptimizer(
            MOI.Utilities.UniversalFallback(MOI.Utilities.Model{Float64}()),
            model,
        )
    end
    MOI.set(model, MOI.Silent(), true)
    MOI.Test.runtests(
      model,
      MOI.Test.Config(
          rtol = 1e-3,
          atol = 1e-3,
          exclude = Any[
              MOI.ConstraintBasisStatus,
              MOI.VariableBasisStatus,
              MOI.ObjectiveBound,
              MOI.SolverVersion,
          ],
      ),
      include = Regex[
          r"test_variable_solve_with_lowerbound$",
          r"test_modification_delete_variables_in_a_batch$",
      ],
    )
end

@testset "$cache" for cache in [false, true]
   tests(cache)
end
```
produces
```
Test Summary: | Pass  Total  Time
false         |   18     18  0.9s
test_modification_delete_variables_in_a_batch: Test Failed at /home/blegat/.julia/packages/MathOptInterface/vrZOD/src/Test/test_modification.jl:827
  Expression: isapprox(MOI.get(model, MOI.ObjectiveValue()), T(2), config)
   Evaluated: isapprox(6.0, 2.0, MathOptInterface.Test.Config{Float64}(0.001, 0.001, MathOptInterface.OPTIMAL, MathOptInterface.INFEASIBLE, Any[MathOptInterface.ConstraintBasisStatus, MathOptInterface.VariableBasisStatus, MathOptInterface.ObjectiveBound, MathOptInterface.SolverVersion]))

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] test_modification_delete_variables_in_a_batch(model::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{CSDP.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}, config::MathOptInterface.Test.Config{Float64})
   @ MathOptInterface.Test ~/.julia/packages/MathOptInterface/vrZOD/src/Test/test_modification.jl:827
 [3] macro expansion
   @ ~/.julia/packages/MathOptInterface/vrZOD/src/Test/Test.jl:272 [inlined]
 [4] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [5] runtests(model::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{CSDP.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}, config::MathOptInterface.Test.Config{Float64}; include::Vector{Regex}, exclude::Vector{String}, warn_unsupported::Bool, verbose::Bool, exclude_tests_after::VersionNumber)
   @ MathOptInterface.Test ~/.julia/packages/MathOptInterface/vrZOD/src/Test/Test.jl:267
test_variable_solve_with_lowerbound: Test Failed at /home/blegat/.julia/packages/MathOptInterface/vrZOD/src/Test/Test.jl:462
  Expression: isapprox(MOI.get(model, MOI.ConstraintPrimal(), index), solution_value, config)
   Evaluated: isapprox(9.973062820023415e-9, 1.0, MathOptInterface.Test.Config{Float64}(0.001, 0.001, MathOptInterface.OPTIMAL, MathOptInterface.INFEASIBLE, Any[MathOptInterface.ConstraintBasisStatus, MathOptInterface.VariableBasisStatus, MathOptInterface.ObjectiveBound, MathOptInterface.SolverVersion]))

Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] _test_model_solution(model::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{CSDP.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}, config::MathOptInterface.Test.Config{Float64}; objective_value::Float64, variable_primal::Vector{Tuple{MathOptInterface.VariableIndex, Float64}}, constraint_primal::Vector{Tuple{MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex}, Float64}}, constraint_dual::Vector{Tuple{MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex}, Float64}})
   @ MathOptInterface.Test ~/.julia/packages/MathOptInterface/vrZOD/src/Test/Test.jl:462
 [3] _test_model_solution
   @ ~/.julia/packages/MathOptInterface/vrZOD/src/Test/Test.jl:427 [inlined]
 [4] test_variable_solve_with_lowerbound(model::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{CSDP.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}, config::MathOptInterface.Test.Config{Float64})
   @ MathOptInterface.Test ~/.julia/packages/MathOptInterface/vrZOD/src/Test/test_variable.jl:244
 [5] macro expansion
   @ ~/.julia/packages/MathOptInterface/vrZOD/src/Test/Test.jl:272 [inlined]
 [6] macro expansion
   @ ~/.julia/juliaup/julia-1.10.0+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [7] runtests(model::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{CSDP.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}, config::MathOptInterface.Test.Config{Float64}; include::Vector{Regex}, exclude::Vector{String}, warn_unsupported::Bool, verbose::Bool, exclude_tests_after::VersionNumber)
   @ MathOptInterface.Test ~/.julia/packages/MathOptInterface/vrZOD/src/Test/Test.jl:267
Test Summary:                                   | Pass  Fail  Total  Time
true                                            |   16     2     18  0.7s
  test_modification_delete_variables_in_a_batch |    7     1      8  0.7s
  test_variable_solve_with_lowerbound           |    9     1     10  0.0s
ERROR: LoadError: Some tests did not pass: 16 passed, 2 failed, 0 errored, 0 broken.
in expression starting at /home/blegat/.julia/dev/CSDP/test/bug2.jl:33
```

Here's a reproducible example:
```julia
julia> import MathOptInterface as MOI

julia> import CSDP

julia> inner = MOI.instantiate(CSDP.Optimizer; with_bridge_type = Float64)
MOIB.LazyBridgeOptimizer{MOIU.CachingOptimizer{CSDP.Optimizer, MOIU.UniversalFallback{MOIU.Model{Float64}}}}
with 0 variable bridges
with 0 constraint bridges
with 0 objective bridges
with inner model MOIU.CachingOptimizer{CSDP.Optimizer, MOIU.UniversalFallback{MOIU.Model{Float64}}}
  in state EMPTY_OPTIMIZER
  in mode AUTOMATIC
  with model cache MOIU.UniversalFallback{MOIU.Model{Float64}}
    fallback for MOIU.Model{Float64}
  with optimizer CSDP.Optimizer

julia> MOI.set(inner, MOI.Silent(), true)

julia> model = MOI.Utilities.CachingOptimizer(MOI.Utilities.Model{Float64}(), inner)
MOIU.CachingOptimizer{MOIB.LazyBridgeOptimizer{MOIU.CachingOptimizer{CSDP.Optimizer, MOIU.UniversalFallback{MOIU.Model{Float64}}}}, MOIU.Model{Float64}}
in state EMPTY_OPTIMIZER
in mode AUTOMATIC
with model cache MOIU.Model{Float64}
with optimizer MOIB.LazyBridgeOptimizer{MOIU.CachingOptimizer{CSDP.Optimizer, MOIU.UniversalFallback{MOIU.Model{Float64}}}}
  with 0 variable bridges
  with 0 constraint bridges
  with 0 objective bridges
  with inner model MOIU.CachingOptimizer{CSDP.Optimizer, MOIU.UniversalFallback{MOIU.Model{Float64}}}
    in state EMPTY_OPTIMIZER
    in mode AUTOMATIC
    with model cache MOIU.UniversalFallback{MOIU.Model{Float64}}
      with 1 optimizer attribute
      fallback for MOIU.Model{Float64}
    with optimizer CSDP.Optimizer

julia> x = MOI.add_variables(model, 2)
2-element Vector{MathOptInterface.VariableIndex}:
 MOI.VariableIndex(1)
 MOI.VariableIndex(2)

julia> MOI.set(model, MOI.ObjectiveSense(), MOI.MIN_SENSE)

julia> f = 1.0 * x[1] + 2.0 * x[2]
0.0 + 1.0 MOI.VariableIndex(1) + 2.0 MOI.VariableIndex(2)

julia> MOI.set(model, MOI.ObjectiveFunction{typeof(f)}(), f)

julia> MOI.add_constraint.(model, x, MOI.GreaterThan(1.0))
2-element Vector{MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.GreaterThan{Float64}}}:
 MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.GreaterThan{Float64}}(1)
 MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.GreaterThan{Float64}}(2)

julia> MOI.optimize!(model)

julia> MOI.delete(model, x[1])

julia> MOI.optimize!(model)

julia> print(inner)
Minimize ScalarAffineFunction{Float64}:
 1.0 + 2.0 v[-2]

Subject to:

VariableIndex-in-GreaterThan{Float64}
 v[-2] >= 1.0

julia> @assert isapprox(MOI.get(model, MOI.ObjectiveValue()), 2.0)
ERROR: AssertionError: isapprox(MOI.get(model, MOI.ObjectiveValue()), 2.0)
Stacktrace:
 [1] top-level scope
   @ REPL[283]:1
```

Note that the objective of `inner` is `1.0 + 2.0 v[-2]` not `0.0 + 2.0 v[-2]`.

--- a/src/Bridges/bridge_optimizer.jl
+++ b/src/Bridges/bridge_optimizer.jl
@@ -1507,8 +1507,9 @@ function MOI.get(
     elseif !Variable.has_bridges(Variable.bridges(b))
         return MOI.get(b.model, attr, ci)
     else
-        msg = "unable to get $attr because it might contain bridged variables"
-        throw(MOI.GetAttributeNotAllowed(attr, msg))
+        # Get the fallback, because the function might contain bridged
+        # variables.
+        return MOI.Utilities.get_fallback(b, attr, ci)
     end
 end
 

--- a/src/Bridges/bridge_optimizer.jl
+++ b/src/Bridges/bridge_optimizer.jl
@@ -1499,18 +1499,26 @@ end
 function MOI.get(
     b::AbstractBridgeOptimizer,
     attr::MOI.ConstraintPrimal,
-    ci::MOI.ConstraintIndex,
-)
+    ci::MOI.ConstraintIndex{F,S},
+) where {F<:MOI.AbstractScalarFunction,S<:MOI.AbstractScalarSet}
     if is_bridged(b, ci)
         MOI.throw_if_not_valid(b, ci)
         return call_in_context(MOI.get, b, ci, attr)
-    elseif !Variable.has_bridges(Variable.bridges(b))
-        return MOI.get(b.model, attr, ci)
-    else
-        # Get the fallback, because the function might contain bridged
-        # variables.
+    end
+    if Variable.has_bridges(Variable.bridges(b))
+        # In this case, the scalar constraint might contain bridged variables
+        # that include constants, like `x >= 1` being mapped to `y >= 0`,
+        # `x := y + 1`. If this is true, then `normalize_and_add_constraint`
+        # may move the constants into the set, and querying `ConstraintPrimal`
+        # from `b.model` will return the value of `y`, not `y + 1`. As a
+        # work-around, we use `get_fallback`, which first queries
+        # `ConstraintFunction` and then evaluates the function at the point
+        # defined by `VariablePrimal`. Querying `ConstraintFunction` accounts
+        # for the case where constants were moved to the set, so we return the
+        # correct value.
         return MOI.Utilities.get_fallback(b, attr, ci)
     end
+    return MOI.get(b.model, attr, ci)
 end
 
 function MOI.supports(

--- a/src/Bridges/bridge_optimizer.jl
+++ b/src/Bridges/bridge_optimizer.jl
@@ -1496,6 +1496,22 @@ function MOI.get(
     return unbridged_function(b, func)
 end
 
+function MOI.get(
+    b::AbstractBridgeOptimizer,
+    attr::MOI.ConstraintPrimal,
+    ci::MOI.ConstraintIndex,
+)
+    if is_bridged(b, ci)
+        MOI.throw_if_not_valid(b, ci)
+        return call_in_context(MOI.get, b, ci, attr)
+    elseif !Variable.has_bridges(Variable.bridges(b))
+        return MOI.get(b.model, attr, ci)
+    else
+        msg = "unable to get $attr because it might contain bridged variables"
+        throw(MOI.GetAttributeNotAllowed(attr, msg))
+    end
+end
+
 function MOI.supports(
     b::AbstractBridgeOptimizer,
     attr::MOI.AbstractConstraintAttribute,

--- a/test/Bridges/lazy_bridge_optimizer.jl
+++ b/test/Bridges/lazy_bridge_optimizer.jl
@@ -333,6 +333,24 @@ function test_MOI_runtests_GeometricSDPAModel()
     return
 end
 
+function test_get_primal_variable_bridge()
+    cache = MOI.Utilities.Model{Float64}()
+    x, c1 = MOI.add_constrained_variable(cache, MOI.GreaterThan(1.0))
+    c2 = MOI.add_constraint(cache, 1.0x, MOI.EqualTo(2.0))
+    inner = MOI.Utilities.MockOptimizer(StandardSDPAModel{Float64}())
+    model = MOI.Bridges.full_bridge_optimizer(inner, Float64)
+    index_map = MOI.copy_to(model, cache)
+    MOI.Utilities.set_mock_optimize!(
+        inner,
+        mock -> MOI.Utilities.mock_optimize!(mock, [1.0]),
+    )
+    MOI.optimize!(model)
+    @test MOI.get(model, MOI.VariablePrimal(), index_map[x]) == 2.0
+    @test MOI.get(model, MOI.ConstraintPrimal(), index_map[c1]) == 2.0
+    @test MOI.get(model, MOI.ConstraintPrimal(), index_map[c2]) == 2.0
+    return
+end
+
 function test_index_constraint_conflict()
     optimizer = StandardSDPAModel{Float64}()
     model = MOI.Bridges.full_bridge_optimizer(optimizer, Float64)

